### PR TITLE
Add SASRec example

### PR DESCRIPTION
**Pull request by abheesht17**

### CODE REVIEW

File: examples/sas_rec.py [hunk 1]
```diff
@@ -0,0 +1,657 @@
...
+We then simply use the Keras `model.predict()` method. Under the hood, it calls
+the `BruteForceRetrieval` layer to perform the actual retrieval.
+
+Note that this model can retrieve movies already watched by the user. We could
+easily add logic to remove them if that is desirable.
+"""
+
+for ele in val_ds.unbatch().take(1):
+    test_sample = ele[0]
+    test_sample["item_ids"] = test_sample["item_ids"][tf.newaxis, ...]
```
Comment (hertschuh) [inline examples/sas_rec.py:640]: Wait, what is this doing? Why do you need `tf.newaxis`? Normally you use this in the context where you're adding dimensions. Is this doing `expand_dims(..., axis=0)`?

**Review by hertschuh — COMMENTED** [review]
Thank you for putting this together!

File: examples/sas_rec.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+example, we built a GRU-based sequential retrieval model. In this example, we
+will build a popular Transformer decoder-based model named Self-Attentive
+Sequential Recommendation (SasRec) for the same sequential recommendation task.
+
+Let's begin by importing all the necessary libraries.
+"""
+
+import collections
+import os
+
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply examples/sas_rec.py:18]: We haven't done this in other examples, but we'll have to do it on keras.io so that it is labelled as a Keras 3 example.

File: examples/sas_rec.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+    examples = {
+        "sequence": [],
+        "negative_sequence": [],
+    }
+
+    for user_id in sequences:
+        sequence = [int(d["movie_id"]) for d in sequences[user_id]]
+
+        # Get negative sequence.
+        # TODO: Check if this can be done more efficiently with TF.
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply examples/sas_rec.py]: Not slow

File: examples/sas_rec.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+implies maximising `positive_logits`. Similarly, we want to minimise
+`negative_logits`.
+"""
+
+
+def _sas_rec_kernel_initializer():
+    return keras.initializers.GlorotUniform()
+
+
+class SasRec(keras.Model):
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply examples/sas_rec.py:377]: I think `SasRec` is okay here. Essentially, if an abbreviation can be pronounced (instead of saying the letters out loud), we keep it lowercase. For example, `BertBackbone` instead of `BERTBackbone`: https://github.com/keras-team/keras-hub/blob/master/keras_hub/src/models/bert/bert_backbone.py#L18.

File: examples/sas_rec.py (hunk 1 above)
Comment (abheesht17) [reply examples/sas_rec.py]: Whoops, meant to use `ops.expand_dims`. Need to add a batch dimension for the forward pass

File: examples/sas_rec.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hertschuh) [reply examples/sas_rec.py:377]: Oh you're right, that's what we do for class names.

File: examples/sas_rec.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
+"""
+# Recommending movies: retrieval using a sequential transformer model (SasRec)
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply examples/sas_rec.py:2]: Oh you should change this one too. All the other ones are good now.

File: examples/sas_rec.py [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+We then simply use the Keras `model.predict()` method. Under the hood, it calls
+the `BruteForceRetrieval` layer to perform the actual retrieval.
+
+Note that this model can retrieve movies already watched by the user. We could
+easily add logic to remove them if that is desirable.
+"""
+
+for ele in val_ds.unbatch().take(1):
+    test_sample = ele[0]
+    test_sample["item_ids"] = ops.expand_dims(test_sample["item_ids"], axis=0)
```
Comment (hertschuh) [inline examples/sas_rec.py:636]: You're in a TF context, here, you'll get tf tensors, you should use `tf.expand_dims`. `ops.expand_dims` might work once we switch the backend to JAX.

**Review by hertschuh — APPROVED** [review]
Thanks for the PR!
